src/main/java/org/prebid/server/bidder/adkerneladn/AdkernelAdnBidder.java

```diff
@@ -0,0 +1,256 @@
+
+                if (compatBanner.getFormat().size() > 1) {
+                    bannerBuilder.format(compatBanner.getFormat().subList(1, compatBanner.getFormat().size()));
+                } else {
+                    bannerBuilder.format(Collections.emptyList());
+                }
+                bannerBuilder
+                        .w(format.getW())
+                        .h(format.getH());
+
+                impBuilder.banner(bannerBuilder.build());
+            }
+        }
+
+        return impBuilder.build();
+    }
+
+    private static List<HttpRequest<BidRequest>> buildAdapterRequests(BidRequest prebidBidrequest,
+                                                                      Map<ExtImpAdkernelAdn, List<Imp>> pubToImps,
+                                                                      String endpointUrl) {
+        final List<HttpRequest<BidRequest>> result = new ArrayList<>();
+
+        for (ExtImpAdkernelAdn impExt : pubToImps.keySet()) {
+            final BidRequest outgoingRequest = createBidRequest(prebidBidrequest, pubToImps.get(impExt));
+            final String body = Json.encode(outgoingRequest);
+            result.add(HttpRequest.of(HttpMethod.POST, buildEndpoint(impExt, endpointUrl), body, headers(),
+                    outgoingRequest));
+        }
+
+        return result;
+    }
+
+    private static BidRequest createBidRequest(BidRequest prebidBidRequest, List<Imp> imps) {
+        final BidRequest.BidRequestBuilder bidRequestBuilder = prebidBidRequest.toBuilder();
+
+        final List<Imp> modifiedImps = new ArrayList<>();
+        for (Imp imp : imps) {
+            final Imp modifiedImp = imp.toBuilder()
+                    .tagid(imp.getId())
+                    .build();
+
+            modifiedImps.add(modifiedImp);
+        }
+        bidRequestBuilder.imp(modifiedImps);
+
+        if (prebidBidRequest.getSite() != null) {
+            final Site modifySite = prebidBidRequest.getSite().toBuilder()
+                    .publisher(null).domain("").build();
+
+            bidRequestBuilder.site(modifySite);
+        }
+        if (prebidBidRequest.getApp() != null) {
+            final App modifyApp = prebidBidRequest.getApp().toBuilder()
+                    .publisher(null).build();
+
+            bidRequestBuilder.app(modifyApp);
+        }
+        return bidRequestBuilder.build();
+    }
+
+    // Builds enpoint url based on adapter-specific pub settings from imp.ext
+    private static String buildEndpoint(ExtImpAdkernelAdn impExt, String endpointUrl) {
+
+        if (impExt.getHost() != null) {
+            try {
+                final URL defaultUrl = new URL(endpointUrl);
+                final String defaultDomain = defaultUrl.getHost();
+                final String defaultPort = defaultUrl.getPort() == -1 ? "" : ":" + defaultUrl.getPort();
+
+                endpointUrl = endpointUrl.replace(defaultDomain + defaultPort, impExt.getHost());
```

Comment:
It is not a good practice to change input args, suggest to rework this methos like:
```
    private static String buildEndpoint(ExtImpAdkernelAdn impExt, String endpointUrl) {
        final String updatedEndpointUrl;

        if (impExt.getHost() != null) {
            final URL url;
            try {
                url = new URL(endpointUrl);
            } catch (MalformedURLException e) {
                throw new PreBidException(
                        String.format("Error occurred while parsing AdkernelAdn endpoint url: %s", endpointUrl), e);
            }

            final String currentHostAndPort = url.getHost() + (url.getPort() == -1 ? "" : ":" + url.getPort());
            updatedEndpointUrl = endpointUrl.replace(currentHostAndPort, impExt.getHost());
        } else {
            updatedEndpointUrl = endpointUrl;
        }

        return String.format("%s%s", updatedEndpointUrl, impExt.getPubId());
    }
```